types/validation.go

```diff
@@ -427,3 +468,113 @@ func verifyBasicValsAndCommit(vals *ValidatorSet, commit *Commit, height int64,
 
 	return nil
 }
+
+func verifyAggregatedCommit(
+	chainID string,
+	vals *ValidatorSet,
+	commit *Commit,
+	votingPowerNeeded int64,
+	ignoreSig func(CommitSig) bool,
+	countSig func(CommitSig) bool,
+	lookUpByIndex bool,
+) error {
+	var (
+		val                *Validator
+		valIdx             int32
+		seenVals           = make(map[int32]int, len(commit.Signatures))
+		talliedVotingPower int64
+		aggSig1, aggSig2   []byte
+		msg1, msg2         []byte
+	)
+
+	pubkeys1 := make([]*bls12381.PubKey, 0, len(commit.Signatures))
+	pubkeys2 := make([]*bls12381.PubKey, 0, len(commit.Signatures))
+
+	for idx, commitSig := range commit.Signatures {
+		// skip over signatures that should be ignored
+		if ignoreSig(commitSig) {
+			continue
+		}
+
+		// If the vals and commit have a 1-to-1 correspondence we can retrieve
+		// them by index else we need to retrieve them by address
+		if lookUpByIndex {
+			val = vals.Validators[idx]
+		} else {
+			valIdx, val = vals.GetByAddressMut(commitSig.ValidatorAddress)
+
+			// if the signature doesn't belong to anyone in the validator set
+			// then we just skip over it
+			if val == nil {
+				continue
+			}
+
+			// because we are getting validators by address we need to make sure
+			// that the same validator doesn't commit twice
+			if firstIndex, ok := seenVals[valIdx]; ok {
+				secondIndex := idx
+				return fmt.Errorf("double vote from %v (%d and %d)", val, firstIndex, secondIndex)
+			}
+			seenVals[valIdx] = idx
+		}
+
+		if val.PubKey == nil {
+			return fmt.Errorf("validator %v has a nil PubKey at index %d", val, idx)
+		}
+
+		if commitSig.BlockIDFlag == BlockIDFlagAggCommit || commitSig.BlockIDFlag == BlockIDFlagAggCommitAbsent {
+			if aggSig1 == nil {
+				aggSig1 = commitSig.Signature
+				msg1 = commit.VoteSignBytes(chainID, int32(idx))
+			}
```

Comment:
```suggestion
			if commitSig.BlockIDFlag == BlockIDFlagAggCommit {
				if aggSig1 != nil {
					return fmt.Errorf("duplicate aggregate signature in validator %v at index %d", val, idx)
				}
				aggSig1 = commitSig.Signature
				msg1 = commit.VoteSignBytes(chainID, int32(idx))
```

Reply:
We need to detect a malformed commit with two `BlockIDFlagAggCommit` entries

Reply:
Also, I'd add a second `if` (maybe turn the two ifs into a case) if we want the `BlockIDFlagAggCommit` to always appear before all the `BlockIDFlagAggCommitAbsent` (but that's not a must).

If you do it here, please do it as well below (for `nil` signature).